internal/repository/master_index.go

```diff
@@ -266,47 +266,68 @@ func (mi *MasterIndex) MergeFinalIndexes() {
 
 // RebuildIndex combines all known indexes to a new index, leaving out any
 // packs whose ID is contained in packBlacklist. The new index contains the IDs
-// of all known indexes in the "supersedes" field.
-func (mi *MasterIndex) RebuildIndex(packBlacklist restic.IDSet) (*Index, error) {
+// of all known indexes in the "supersedes" field. The IDs are also returned in
+// the IDSet obsolete
+func (mi *MasterIndex) RebuildIndex(ctx context.Context, repo restic.Repository, packBlacklist restic.IDSet) (obsolete restic.IDSet, err error) {
 	mi.idxMutex.Lock()
 	defer mi.idxMutex.Unlock()
 
 	debug.Log("start rebuilding index of %d indexes, pack blacklist: %v", len(mi.idx), packBlacklist)
 
 	newIndex := NewIndex()
+	obsolete = restic.NewIDSet()
+	newIdx := mi.idx[:0]
 
-	ctx, cancel := context.WithCancel(context.TODO())
-	defer cancel()
+	finalize := func() error {
+		newIndex.Finalize()
+		if _, err := SaveIndex(ctx, repo, newIndex); err != nil {
+			return err
+		}
+		newIdx = append(newIdx, newIndex)
+		newIndex = NewIndex()
+		return nil
+	}
 
 	for i, idx := range mi.idx {
-		debug.Log("adding index %d", i)
-
-		for pb := range idx.Each(ctx) {
-			if packBlacklist.Has(pb.PackID) {
-				continue
+		if idx.Final() {
+			ids, err := idx.IDs()
+			if err != nil {
+				debug.Log("index %d does not have an ID: %v", err)
+				return nil, err
 			}
 
-			newIndex.Store(pb)
-		}
+			debug.Log("adding index ids %v to supersedes field", ids)
 
-		if !idx.Final() {
+			err = newIndex.AddToSupersedes(ids...)
```

Comment:
I'm not sure whether this is intentional, but obsolete is only set on a few select indexes (the first one derive from each merged index).

Reply:
Yes, this was intentional. I just thought about what would happen if the repacking was canceled. Even though the `supersedes` field is not really used at the moment, I didn't want to have new index files saved with a `supersedes` field set to an index file that has not been fully processed. Of course merging the indexes to one single big one doesn't improve the situation as then all IDs will be written as `supersedes` after all index entries are processed.